### PR TITLE
[TS migration] Migrate 'BlockingViews' component to TypeScript

### PR DESCRIPTION
**Pull request by cdOut**
### Fixed Issues

$ https://github.com/Expensify/App/issues/25133

### Tests

1. Log into the app, click through main flows to see whether they are performing as expected, since components migrated here are used all throughout the app as a wrapper to other screens / modals.
2. FullPageNotFoundView can be checked when going into Settings > Profile, that view is based on it.
3. FullPageOfflineBlockingView can be checked when going into Settings > Wallet and pressing Enable wallet.
4. BlockingView is part of the two others so if these work then this base component also works.

- [x] Verify that no errors appear in the JS console

### QA Steps

1. Log into the app, click through main flows to see whether they are performing as expected, since components migrated here are used all throughout the app as a wrapper to other screens / modals.
2. FullPageNotFoundView can be checked when going into Settings > Profile, that view is based on it.
3. FullPageOfflineBlockingView can be checked when going into Settings > Wallet and pressing Enable wallet.
4. BlockingView is part of the two others so if these work then this base component also works.

- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/Expensify/App/assets/88325488/788f7de8-dbe5-4efe-a0cd-41d6013ce1d2

</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/Expensify/App/assets/88325488/0014e12d-7695-46ab-9cbd-1c2c185b6f15

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/Expensify/App/assets/88325488/ce3b10c6-0d24-409e-aa1c-677b34558aa6

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/Expensify/App/assets/88325488/6aca1e83-2ee0-481f-abbe-03c011b67419

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/Expensify/App/assets/88325488/c21f1162-53dd-481c-a944-aadfd4d5c130

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/Expensify/App/assets/88325488/3e1dd203-db68-45c5-a78a-a28a25e6e54d

</details>

### CODE REVIEW

**Comment by cdOut** [comment]
Text and TextLink are here temporarily, they were migrated in another issue. Will remove them after the referenced issue PR gets merged.

File: src/components/BlockingViews/FullPageOfflineBlockingView.tsx [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+import useLocalize from '@hooks/useLocalize';
+import useNetwork from '@hooks/useNetwork';
+import BlockingView from './BlockingView';
+
+type FullPageOfflineBlockingViewProps = {
+    /** Child elements */
+    children: React.ReactNode;
+};
+
+function FullPageOfflineBlockingView({children}: FullPageOfflineBlockingViewProps) {
```
Comment (blazejkustra) [inline src/components/BlockingViews/FullPageOfflineBlockingView.tsx:12]:
```suggestion
import ChildrenProps from '@src/types/utils/ChildrenProps';

function FullPageOfflineBlockingView({children}: ChildrenProps) {
```

File: src/components/TextLink.tsx [hunk 2]
```diff
@@ -0,0 +1,77 @@
+import React, {ForwardedRef, forwardRef, KeyboardEventHandler, MouseEventHandler} from 'react';
```
Comment (blazejkustra) [inline src/components/TextLink.tsx:1]: Remove this before proceeding with cross review

File: src/components/BlockingViews/FullPageNotFoundView.tsx [hunk 3]
```diff
@@ -7,52 +6,51 @@ import useLocalize from '@hooks/useLocalize';
...
+    subtitleKey = 'notFound.pageNotFound',
+    linkKey = 'notFound.goBackHome',
+    onBackButtonPress = () => Navigation.goBack(ROUTES.HOME),
+    shouldShowLink = true,
+    shouldShowBackButton = true,
+    onLinkPress = () => Navigation.dismissModal,
+}: FullPageNotFoundViewProps) {
     const styles = useThemeStyles();
     const {translate} = useLocalize();
     if (shouldShow) {
```
Comment (blazejkustra) [inline src/components/BlockingViews/FullPageNotFoundView.tsx:56]: Nit:
```suggestion

    if (shouldShow) {
```

**Review by blazejkustra — CHANGES_REQUESTED** [review]

File: src/components/Text.tsx [hunk 4]
```diff
@@ -1,28 +1,29 @@
 import React, {ForwardedRef} from 'react';
-// eslint-disable-next-line no-restricted-imports
```
[the remark this answers is not in the document]
Comment (cdOut) [reply src/components/Text.tsx:2]: This is part of the TextLink issue, I don't think I should be changing it before that one gets merged

**Comment by cdOut** [comment]
> Text and TextLink are here temporarily, they were migrated in another issue. Will remove them after the referenced issue PR gets merged.

All files related to the TS migration of TextLink were migrated from main since the blocking issue was merged.

**Review by blazejkustra — APPROVED** [review]
LGTM, please update tests and QA sections in the description 😄

File: src/components/BlockingViews/BlockingView.tsx [hunk 5]
```diff
@@ -10,51 +10,52 @@ import Navigation from '@libs/Navigation/Navigation';
...
+function BlockingView({
+    icon,
+    iconColor,
+    title,
+    subtitle = '',
+    linkKey = 'notFound.goBackHome',
+    shouldShowLink = false,
+    iconWidth = variables.iconSizeSuperLarge,
+    iconHeight = variables.iconSizeSuperLarge,
+    onLinkPress = () => Navigation.dismissModal,
```
Comment (VickyStash) [inline src/components/BlockingViews/BlockingView.tsx:56]: Why have you removed the function call?
```suggestion
    onLinkPress = () => Navigation.dismissModal(),
```

File: src/components/BlockingViews/BlockingView.tsx [hunk 6]
```diff
@@ -10,51 +10,52 @@ import Navigation from '@libs/Navigation/Navigation';
...
+    shouldEmbedLinkWithSubtitle,
```
Comment (VickyStash) [inline src/components/BlockingViews/BlockingView.tsx:57]:
```suggestion
    shouldEmbedLinkWithSubtitle = false,
```

File: src/components/BlockingViews/FullPageNotFoundView.tsx [hunk 7]
```diff
@@ -7,54 +6,54 @@ import useLocalize from '@hooks/useLocalize';
...
-    subtitleKey: 'notFound.pageNotFound',
-    linkKey: 'notFound.goBackHome',
-    onBackButtonPress: () => Navigation.goBack(ROUTES.HOME),
-    shouldShowLink: true,
-    shouldShowBackButton: true,
-    onLinkPress: () => Navigation.dismissModal(),
+    onLinkPress: () => void;
 };
 
 // eslint-disable-next-line rulesdir/no-negated-variables
```
Comment (VickyStash) [inline src/components/BlockingViews/FullPageNotFoundView.tsx:42]: Do we still need this disable?

File: src/components/BlockingViews/FullPageNotFoundView.tsx [hunk 8]
```diff
@@ -7,54 +6,54 @@ import useLocalize from '@hooks/useLocalize';
...
+function FullPageNotFoundView({
+    children = null,
+    shouldShow = false,
+    titleKey = 'notFound.notHere',
+    subtitleKey = 'notFound.pageNotFound',
+    linkKey = 'notFound.goBackHome',
+    onBackButtonPress = () => Navigation.goBack(ROUTES.HOME),
+    shouldShowLink = true,
+    shouldShowBackButton = true,
+    onLinkPress = () => Navigation.dismissModal,
```
Comment (VickyStash) [inline src/components/BlockingViews/FullPageNotFoundView.tsx:52]: Why have you removed the function call?
```suggestion
    onLinkPress = () => Navigation.dismissModal(),
```

File: src/components/BlockingViews/FullPageNotFoundView.tsx (hunk 7 above)
Comment (cdOut) [reply src/components/BlockingViews/FullPageNotFoundView.tsx:42]: Yes we do, this is because of the name of this component which I will not be changing.

**Review by VickyStash — APPROVED** [review]
LGTM 👍, please don't forget to add test steps/screenshots

**Comment by cdOut** [comment]
cc: @0xmiroslav

**Comment by cdOut** [comment]
@0xmiroslav merged and fixed conflicts

**Review by 0xmiroslav — APPROVED** [review]

**Review by 0xmiroslav — APPROVED** [review]
@joelbettner all yours

**Review by joelbettner — APPROVED** [review]
👍 LGTM

**Pull request merged**